src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs

```diff
@@ -3239,6 +4648,112 @@ ref Unsafe.As<T, long>(ref dst),
             SpanHelpers.Replace(ref src, ref dst, oldValue, newValue, length);
         }
 
+        /// <summary>
+        /// Copies <paramref name="source"/> to <paramref name="destination"/>, replacing all occurrences of <paramref name="oldValue"/> with <paramref name="newValue"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of the elements in the spans.</typeparam>
+        /// <param name="source">The span to copy.</param>
+        /// <param name="destination">The span into which the copied and replaced values should be written.</param>
+        /// <param name="oldValue">The value to be replaced with <paramref name="newValue"/>.</param>
+        /// <param name="newValue">The value to replace all occurrences of <paramref name="oldValue"/>.</param>
+        /// <param name="comparer">The <see cref="IEqualityComparer{T}"/> implementation to use when comparing elements, or <see langword="null"/> to use the default <see cref="IEqualityComparer{T}"/> for the type of an element.</param>
+        /// <exception cref="ArgumentException">The <paramref name="destination"/> span was shorter than the <paramref name="source"/> span.</exception>
+        /// <exception cref="ArgumentException">The <paramref name="source"/> and <paramref name="destination"/> were overlapping but not referring to the same starting location.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe void Replace<T>(this ReadOnlySpan<T> source, Span<T> destination, T oldValue, T newValue, IEqualityComparer<T>? comparer = null)
+        {
+            nuint length = (uint)source.Length;
+            if (length == 0)
+            {
+                return;
+            }
+
+            if (length > (uint)destination.Length)
+            {
+                ThrowHelper.ThrowArgumentException_DestinationTooShort();
+            }
+
+            ref T src = ref MemoryMarshal.GetReference(source);
+            ref T dst = ref MemoryMarshal.GetReference(destination);
```

Comment:
Potential perf improvement: You may want to have something like this:

```cs
if (Unsafe.AreSame(ref src, ref dst)) { return Replace(source, oldValue, newValue, comparer); }
```

The reason is that later in this method, _every_ element inspection incurs a write back to `dst`. If `T` contains a reference type, these writes call a GC helper, which could make the aggregate operation quite expensive. The in-place method only writes back if the value has changed, minimizing invocations to this GC helper.

Reply:
It could... though I don't know how common it is to call Replace(source, source, old, new) when you could have just done Replace(source, old, new). We'd be adding the extra comparison for what feels like a corner case?